### ISSUE TITLE
Possible infinite recursion between pagination-handling and 404-page-generation with bad URL-parameter

### ISSUE THREAD

**Issue by trent2** [contao/core#8060]
Using modules which offer pagination on a 404-page might result in a condition where functions call each other repeatedly.
E.g. Having an event list (module XXX) with "items per page" != 0 and using `&page_eXXX`(some number outside the range of valid pages) as an URL-parameter triggers a 404.
If the same module XXX appears on the 404-page it again receives the parameter `page_eXXX` and triggers the 404 again. In the end this results in some kind of "stack overflow"/maximum memory exhausted error. On my site this was triggered by some search machine bot which lead to clogging the system log with 404-errors, resulting in a 1GB-log table in a relatively short period of time.
This can be reproduces in the demo by
* using a "real" 404-page (not a 302-redirection as in the demo)
* change "items per page" to some value != 0 and "number of events" to 0, e.g. in module 35.
* use `page_e35`=invalid page number as an URL-parameter

I don't like the remedy not to use paginated modules on a 404-page, why shouldn't I? Maybe the URL-parameter should not be forwarded to the 404-page, maybe a 404 should not be generated at all?

**Comment by fritzmg** [comment]
Link to forum discussion for reference: https://community.contao.org/de/showthread.php?59440-PHP-st%FCzt-ab-memory-size-exhausted-Log-File-wird-zugem%FCllt

**Comment by fritzmg** [comment]
Only solution I can think of right now:
```php
if ($objPage->type != 'error_404')
{
	/** @var \PageError404 $objHandler */
	$objHandler = new $GLOBALS['TL_PTY']['error_404']();
	$objHandler->generate($objPage->id);
}
```
But that's not really that elegant.

**Comment by leofeyer** [comment]
@contao/developers /cc

**Comment by aschempp** [comment]
As far as I can understand, `$objPage->type` would not be `error_404` in that case, it would still be the original page. The most simple solution would be for the event module to unset the page parameter before throwing the out-of-bounds exception?

**Comment by trent2** [comment]
The second time the if statement is hit, `$objPage->type==error_404` so this basically works (tried it). IMHO it also makes sense not to generate a 404 when the page type already is a 404.

**Comment by aschempp** [comment]
So basically that case could be handled directly in the error404 page type?

**Comment by fritzmg** [comment]
> The second time the if statement is hit, $objPage->type=='error_404' so this basically works (tried it). IMHO it also makes sense not to generate a 404 when the page type already is a 404.

Indeed, because https://github.com/contao/core/blob/3.5.4/system/modules/core/pages/PageError404.php#L96 overwrites the global page object with the new 404 page:
```php
$objPage = $obj404->loadDetails();
```

> So basically that case could be handled directly in the error404 page type?

I guess so. In the beginning of `PageError404::generate()` you could simple check if the current global page object is already of the same type and return without doing anything.

**Comment by aschempp** [comment]
> In the beginning of PageError404::generate() you could simply check if the current global page object is already of the same type and return without doing anything.

Thats what I meant. Thinking about it, it wouldn't help though. Because returning would continue to render the event list with invalid page parameter, and that's not what we want.

**Comment by trent2** [comment]
Isn't the question really what output do we expect from a module on an error page that was generated *in lack of an appropriate output* of that same module?
In the case of the event list using fritzmg's solution just doesn't show any events ( "Aktuell sind keine Termine vorhanden." in German). On the other hand, removing the parameter from the URL would show page one of the event list. Both solutions don't feel "right".

**Comment by asaage** [comment]
Why not set the 404page in htaccess (which is recomended anyway) and point it to a regular page?

**Comment by fritzmg** [comment]
> Why not set the 404page in htaccess (which is recomended anyway) and point it to a regular page?
> That should work at least in singledomain-mode.

That only covers URLs that are not processed by Contao.

**Comment by fritzmg** [comment]
> In the case of the event list using fritzmg's solution just doesn't show any events ( "Aktuell sind keine Termine vorhanden." in German). On the other hand, removing the parameter from the URL would show page one of the event list. Both solutions don't feel "right".

imho the latter solution feels more right

**Comment by trent2** [comment]
> imho the latter solution feels more right

But if one defines some meaningful reaction to the parameter why then have a 404 in the first place? Is it necessary for search engine bots to prevent them from indexing `page_e42=50`, `page_e42=51`, ...? In [this forum post Znrl](https://community.contao.org/de/showthread.php?59440-PHP-st%FCzt-ab-memory-size-exhausted-Log-File-wird-zugem%FCllt&p=387486&viewfull=1#post387486) suggests not to have a 404 but a pagination error template - would that be a suitable compromise?

**Comment by Znrl** [comment]
To be a bit more specific:
For me it doesn't make a real difference if you call:
- page 1 of 0 (no events to show, no parameter in this case)
- page 5 of 4 (not enough events to show)
But in the first case there is no 404 but a (error)message (no events to show) while in the second case there is a 404.
Thats why I would suggest a comparable message including paginations links to get to an existing pagination site.

The site actually exists -> no 404!
And if you are searching for events 50-60 (page 5) while there are only 45 events it leads to an empty result -> message.
Thats at least my opinion.

**Comment by leofeyer** [comment]
Fixed in d63d87bffaa861a9e39f6ebd7032c30dff73e5fb.
